app/src/main/java/ai/elimu/appstore/synchronization/DownloadApplicationAsyncTask.java

```diff
@@ -190,6 +193,13 @@ protected Integer doInBackground(ApplicationVersion... applicationVersions) {
                         Timber.i(e, "IOException");
                     }
                 }
+
+                /**
+                 * Delete downloaded APK file in case its checksum is invalid
+                 */
+                if (!applicationVersion.getChecksumMd5().equals(downloadedApkChecksum)) {
```

Comment:
Won't this cause a NullPointerException if a previously downloaded ApplicationVersion does not have an MD5 checksum? Can we perform the check in reverse order to avoid the NullPointerException?

Reply:
Due to the current implementation of `calculateMd5(File updateFile)`, `downloadedApkChecksum` can also be null.
So the chance is 50-50.
What if we replace null returned value in `calculateMd5(File updateFile)` by empty string and make the call reversed order?

Reply:
> Due to the current implementation of calculateMd5(File updateFile), downloadedApkChecksum can also be null.

Why can the MD5 checksum of a file be null?

Reply:
According to the current implementation, there are 2 cases that we are retunring null value for the checksum
1. 
try {
            digest = MessageDigest.getInstance("MD5");
        } catch (NoSuchAlgorithmException e) {
            Timber.e(e);
            return null;
        }

2. 
try {
            is = new FileInputStream(updateFile);
        } catch (FileNotFoundException e) {
            Timber.e(e);
            return null;
        }

We can replace null value by empty string value and perform the equals check in reversed order.

Reply:
The implementation depends on what you prefer (empty string + equals check VS. null + NullPointerException and then equals check). More important is the decision, what should happen, if no MD5 checksum can be calculated. With the current implementation the file would get deleted. So if for some reason (in the real world, where things are unpredictable), one of those try/catch blocks fails repeatedly, you wouldn't be able to download the APK, even though it could have worked. Suggestion: Don't delete the file in this case, but mark it as "possibly corrupt" or so to the user. But then again, this could lead to users using possibly corrupt APKs :(